src/cluster/websocket/WebSocket.h

```diff
@@ -0,0 +1,301 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#pragma once
+
+#include <list>
+#include <map>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace zeek {
+
+namespace detail {
+
+template<class Proc, class Work>
+class OnLoopProcess;
+}
+
+namespace cluster {
+
+class Backend;
+
+namespace websocket {
+
+
+/**
+ * Library independent interface for a WebSocket client.
+ *
+ * All methods should be safe to be called from Zeek's
+ * main thread, though some may fail if the client has vanished
+ * or vanishes during an operation.
+ */
+class WebSocketClient {
+public:
+    virtual ~WebSocketClient() = default;
+
+    /**
+     * @returns true if the WebSocket client has terminated
+     */
+    virtual bool IsTerminated() const = 0;
+
+    /**
+     * Close the WebSocket connection with the given code/reason.
+     */
+    virtual void Close(uint16_t code = 1000, const std::string& reason = "Normal closure") = 0;
+
+    /**
+     * Information about the send operation.
+     */
+    struct SendInfo {
+        bool success;
+    };
+
+    /**
+     * Thread safe sending.
+     *
+     * This might be called from Zeek's main thread and
+     * must be safe to be called whether or not the connection
+     * with the client is still alive.
+     *
+     * @param sv The buffer to send as a WebSocket message.
+     */
+    virtual SendInfo SendText(std::string_view sv) = 0;
+
+    /**
+     * Send an error in Broker JSON/v1 format to the client.
+     */
+    SendInfo SendError(std::string_view code, std::string_view ctx);
+
+    /**
+     * Send an ACK message Broker JSON/v1 format to the client.
+     */
+    SendInfo SendAck(std::string_view endpoint, std::string_view version);
+
+    /**
+     * @return - has an ACK been sent to the client?
+     */
+    bool IsAcked() const { return acked; }
+
+    /**
+     * @return The WebSocket client's identifier.
+     */
+    virtual const std::string& getId() = 0;
+
+    /**
+     * @return The WebSocket client's remote IP address.
+     */
+    virtual const std::string& getRemoteIp() = 0;
+
+    /**
+     * @return The WebSocket client's remote port.
+     */
+    virtual int getRemotePort() = 0;
+
+    /**
+     * Store the client's subscriptions as "not active".
+     */
+    void SetSubscriptions(const std::vector<std::string>& topic_prefixes);
+
+    /**
+     * @return The client's subscriptions.
+     */
+    const std::vector<std::string>& GetSubscriptions() const;
+
+    /**
+     * Store the client's subscriptions as "not active".
+     */
+    void SetSubscriptionActive(std::string& topic_prefix);
+
+    /**
+     * @return true if all subscriptions have an active status.
+     */
+    bool AllSubscriptionsActive() const;
+
+private:
+    bool acked = false;
+    std::map<std::string, bool> subscriptions_state;
+    std::vector<std::string> subscriptions;
```

Comment:
We only store this value to (marginally?) speed up `GetSubscriptions`. I'd suggest to either just compute it on the fly there, or if really needed introduce a small helper map type which has a constant `keys` function which would internally make sure `subscription_state` and `subscriptions` are consistent (and which would return some set).

My gut feeling is that this is not critical for performance, and if users have a massive number of subscriptions we would first see slowdown elsewhere.